### PR TITLE
Attendance rules

### PR DESCRIPTION
**Pull request by aaronbrowder**
This is a significant new feature that makes the attendance system more flexible. It adds a new "Rules" tab to Attendance where users can create "attendance rules" to alter default behavior for specific people at specific times.

A rule has the following parameters: person, start date, end date, and days of the week. Using a rule, you can (1) automatically apply an absence code each time a new week is generated, (2) override default settings for calculating attendance rate such as min hours and latest allowed arrival time, or (3) make a person exempt from late pickup fees.

The options for (2) and (3) only appear if those features are activated in Settings. However, the Rules tab is always visible because it is IMO a fairly general feature.

As part of this pull request I've also included a significant upgrade to the late fees feature which I made not too long ago.

Unfortunately I was in the middle of this project while you did the big refactor, so I had to merge and it was a bit messy. Everything looks right to me but you might keep an eye out for potential merge issues.

### CODE REVIEW

**Comment by schmave** [comment]
Sounds like a cool feature! Thanks for working through what was most likely a difficult merge. I'll take a look at this within the next week.

File: modelsLibrary/app/models/AttendanceRule.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    rule.setOrganization(org);
+
+    String person_id = form.field("personId").value().get();
+    if (!person_id.isEmpty()) {
+      rule.setPerson(Person.findById(Integer.valueOf(person_id), org));
+    } else {
+      rule.setPerson(null);
+    }
+
+    String latest_start_time = form.field("_latestStartTime").value().get();
```
Comment (schmave) [inline modelsLibrary/app/models/AttendanceRule.java:136]: This line throws an exception if the "enable partial days" setting is false.

File: app/models/LateDepartureGroup.java [hunk 2]
```diff
@@ -9,32 +9,62 @@ public class LateDepartureGroup {
   public String name;
   public List<AttendanceDay> events;
 
-  public Integer lateFee;
+  public Integer late_fee;
+  public Integer late_fee_2;
```
Comment (schmave) [inline app/models/LateDepartureGroup.java:13]: What's the difference between late_fee_2 (and friends) and late_fee?

File: modelsLibrary/app/models/AttendanceRule.java (hunk 1 above)
Comment (aaronbrowder) [reply modelsLibrary/app/models/AttendanceRule.java]: Fixed

File: app/models/LateDepartureGroup.java (hunk 2 above)
Comment (aaronbrowder) [reply app/models/LateDepartureGroup.java:13]: This allows having two late fee rules. For example, at The Open School we charge $6 per 30 minutes after 4:30, then $20 per 20 minutes after 5:30. So late_fee would be 6 and late_fee_2 would be 20.

**Comment by schmave** [comment]
Deployed!

**Comment by aaronbrowder** [comment]
Thanks!

**Pull request merged**